File: client/src/core/evolution.ts
/** Evolution: a card you keep playing grows into its next form. */

import { SPECIES } from "./species";
import { build, byId, type Card } from "./cards";

export const PLAYS_FOR_STAGE: Record<number, number> = { 1: 2, 2: 3, 3: 4 };

/** The source writes some evolutions with an underscore; the key has none. */
const normalise = (name?: string) => name?.replace(/_/g, "");

/** A few chains the source leaves open. */
const EXTRA_EVOLUTION: Record<string, string> = {
  pikachu: "raichu",
};

/** Branching evolution. */
export const BRANCHES: Record<string, string[]> = {
  eevee: ["vaporeon", "jolteon", "flareon", "espeon",
          "umbreon", "leafeon", "glaceon", "sylveon"],
};

/** Three of the eight, not all eight: picking from everything is a menu, picking from three is a decision you can make in a real-time match. */
export const BRANCH_OFFER = 3;

/** Sheets that exist, so we never evolve into something we cannot draw. */
let drawable: (name: string) => boolean = () => true;
export function setDrawableCheck(fn: (name: string) => boolean) {
  drawable = fn;
}

export function branchesFor(species: string): string[] | undefined {
  const list = BRANCHES[species];
  if (!list) return undefined;
  const usable = list.filter((n) => SPECIES[n] && drawable(n));
  return usable.length > 0 ? usable : undefined;
}

/** A random offer of BRANCH_OFFER forms, without repeats. */
export function offerFor(species: string, rng: () => number = Math.random): Card[] | undefined {
  const pool = branchesFor(species);
  if (!pool) return undefined;
  const copy = [...pool];
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  const from = byId(species);
  return copy
    .slice(0, Math.min(BRANCH_OFFER, copy.length))
    .map((n) => build(n, from))
    .filter((c): c is Card => c !== undefined);
}

export function nextOf(species?: string): string | undefined {
  if (!species) return undefined;
  const next = normalise(SPECIES[species]?.evolution) ?? EXTRA_EVOLUTION[species];
  if (!next || !SPECIES[next]) return undefined;
  return next;
}

export function stageOf(species: string): number {
  return SPECIES[species]?.stars ?? 1;
}

/** How many plays this card needs before it changes. undefined means terminal. */
export function playsNeeded(card?: Card): number | undefined {
  if (!card) return undefined;
  if (BRANCHES[card.id]) return PLAYS_FOR_STAGE[stageOf(card.id)];
  if (!nextOf(card.id)) return undefined;
  return PLAYS_FOR_STAGE[stageOf(card.id)] ?? PLAYS_FOR_STAGE[3];
}

/** The evolved card, or undefined if terminal or its sprite is missing. */
export function evolve(card: Card): Card | undefined {
  const next = nextOf(card.id);
  if (!next || !drawable(next)) return undefined;
  return build(next, card);
}

/** Every form this species can reach, itself included. */
export function chainOf(species: string): string[] {
  const out = new Set<string>([species]);
  const walk = (name: string) => {
    if (BRANCHES[name]) for (const b of BRANCHES[name]) out.add(b);
    const next = nextOf(name);
    if (next && !out.has(next)) {
      out.add(next);
      walk(next);
    }
  };
  walk(species);
  return [...out];
}

/** The full evolution line a species sits in, walked in both directions. */
export function lineOf(species: string): string[] {
  const chain = [species];
  let cur = species;
  for (let i = 0; i < 3; i++) {
    const next = nextOf(cur);
    if (!next) break;
    chain.push(next);
    cur = next;
  }
  // Backward until there is no predecessor, not once.
  //
  // A single step was enough for a middle form and wrong for a final one:
  // lineOf("kingambit") returned "bisharp -> kingambit" while
  // lineOf("bisharp") returned the full "pawniard -> bisharp -> kingambit".
  // Callers that reduce a species to the base of its line -- which is what a
  // deck slot is -- got a different answer depending on which member of the
  // chain they happened to ask about.
  //
  // Bounded by the longest chain in the data rather than `while (true)`: the
  // scan is over species that record what they become, and a data error that
  // made two forms point at each other would otherwise hang the caller.
  for (let i = 0; i < 4; i++) {
    const prev = Object.keys(SPECIES).find((n) => nextOf(n) === chain[0]);
    if (!prev || chain.includes(prev)) break;
    chain.unshift(prev);
  }
  return chain;
}

File: server/src/main/java/io/github/excalibase/clashofpokemon/game/rules/Evolution.java
package io.github.excalibase.clashofpokemon.game.rules;

import java.util.List;
import java.util.Map;
import java.util.random.RandomGenerator;

/** What a card becomes, and how many plays it takes. */
public final class Evolution {

  private static final Map<Integer, Integer> PLAYS_FOR_STAGE = Rules.playsForStage();
  private static final Map<String, String> NEXT = Rules.evolutionNext();
  private static final Map<String, Integer> STAGE = Rules.evolutionStage();
  private static final Map<String, List<String>> BRANCHES = Rules.evolutionBranches();

  /** How many of a card's branches are offered at once. */
  public static final int BRANCH_OFFER = Rules.branchOffer();

  private Evolution() {}

  /** The single form this card grows into, or null if it is terminal. */
  public static String nextOf(String species) {
    return species == null ? null : NEXT.get(species);
  }

  public static int stageOf(String species) {
    return STAGE.getOrDefault(species, 1);
  }

  /** How many plays this card needs to grow, or 0 if it never does. */
  public static int playsNeeded(Card card) {
    if (card == null) return 0;
    if (BRANCHES.containsKey(card.id())) {
      return PLAYS_FOR_STAGE.getOrDefault(stageOf(card.id()), 0);
    }
    if (nextOf(card.id()) == null) return 0;
    return PLAYS_FOR_STAGE.getOrDefault(stageOf(card.id()), 0);
  }

  /** Everything this card can branch into, or null if it does not branch. */
  public static List<String> branchesFor(String species) {
    return BRANCHES.get(species);
  }

  /** Three of the branches, drawn at random. */
  public static List<String> offerFor(String species, RandomGenerator rng) {
    List<String> pool = branchesFor(species);
    if (pool == null || pool.isEmpty()) return null;

    List<String> copy = new java.util.ArrayList<>(pool);
    // The same Fisher-Yates the TypeScript uses, in the same direction, so the
    // same seed offers the same three.
    for (int i = copy.size() - 1; i > 0; i--) {
      int j = (int) Math.floor(rng.nextDouble() * (i + 1));
      String swap = copy.get(i);
      copy.set(i, copy.get(j));
      copy.set(j, swap);
    }
    return List.copyOf(copy.subList(0, Math.min(BRANCH_OFFER, copy.size())));
  }

  /** The whole line a card belongs to, from where it stands to the end. */
  public static List<String> chainOf(String species) {
    List<String> out = new java.util.ArrayList<>();
    String at = species;
    while (at != null && !out.contains(at)) {
      out.add(at);
      List<String> branches = branchesFor(at);
      if (branches != null) {
        out.addAll(branches);
        break;
      }
      at = nextOf(at);
    }
    return List.copyOf(out);
  }
}
